### PR TITLE
[linker] Fix comparison of methods wrt generic parameters [#30488]

### PR DESCRIPTION
**Pull request by spouliot**
note: soliciting @jbevain feedback

The check from [1] is fine when it's done once _or_ with an identical
target. However there's nothing to ensure (if done more than once) that
it's being used to represent the same thing every time.

That makes the code from [1] match methods like:

> System.IObservable`1<TResult> System.Reactive.Linq.IQueryLanguage::Repeat(TResult)

and

> System.IObservable`1<TSource> System.Reactive.Linq.QueryLanguage::Repeat(System.IObservable`1<TSource>)

where `TResult` is different in both checks.

This patch tracks what is used as the generic parameter (on first check)
and make sure it's identical if it's used again in the same signature.

Note: this is a very old bug (i.e. it's always been that way) but recent
changes to mono started to report such issues (a good thing) and, for AOT
compilation on XI, that resulted in MT3001 errors.

[1] https://github.com/mono/mono/blob/mono-3.10.0/mcs/tools/linker/Mono.Linker.Steps/TypeMapStep.cs#L250

### CODE REVIEW

File: mcs/tools/linker/Mono.Linker.Steps/TypeMapStep.cs [hunk 1]
```diff
@@ -200,7 +201,10 @@ static bool MethodMatch (MethodDefinition candidate, MethodDefinition method)
 			if (candidate.HasGenericParameters != method.HasGenericParameters)
 				return false;
 
-			if (!TypeMatch (candidate.ReturnType, method.ReturnType))
+			genericPamateters.Clear ();
```
Comment (jbevain) [inline mcs/tools/linker/Mono.Linker.Steps/TypeMapStep.cs]: patameters? :smile:

**Comment by jbevain** [comment]
Thanks for the ping. I'm wondering if you can still get false positives for this, with generic parameters of the same name.

I think the ideal way would be for each method to test if the candidate method is a possible instantiation, but it's bound to be quite hairy.

That being said, it's already much better than what was before :)

File: mcs/tools/linker/Mono.Linker.Steps/TypeMapStep.cs (hunk 1 above)
Comment (spouliot) [inline mcs/tools/linker/Mono.Linker.Steps/TypeMapStep.cs]: The later sounds like the metric system to count patates ;-)

**Comment by spouliot** [comment]
Yeah, I'm not sure it's 100% complete either (in fact I doubt it) but I was more worried about equivalence than the names. I could not see a test case (but I have not spent a lot of time trying either). Right now I'm a bit anxious about how this can affect performance.

I'll go ahead with a variation (avoiding some memory allocations) of this fix - it's minimally a step forward. I'll also add a comment about its possible limits. Maybe we'll need another 6 years to get a test case ;-) but, seriously, with the extra mono runtime checks it should be easier to spot similar cases. Thanks for the feedback :-)

**Pull request closed without merging**